cluster-autoscaler/core/scaledown/actuation/update_latency_tracker_test.go

```diff
@@ -0,0 +1,176 @@
+package actuation
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/taints"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/test"
+)
+
+// mockClock is used to mock time.Now() when testing UpdateLatencyTracker
+// For the n th call to Now() it will return a timestamp after duration[n] to
+// the startTime if n < the length of durations. Otherwise, it will return current time.
+type mockClock struct {
+	startTime time.Time
+	durations []time.Duration
+	index     int
+	mutex     sync.Mutex
+}
+
+// Returns a new NewMockClock object
+func NewMockClock(startTime time.Time, durations []time.Duration) mockClock {
+	return mockClock{
+		startTime: startTime,
+		durations: durations,
+		index:     0,
+	}
+}
+
+// Returns a time after Nth duration from the start time if N < length of durations.
+// Otherwise, returns the current time
+func (m *mockClock) Now() time.Time {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	var timeToSend time.Time
+	if m.index < len(m.durations) {
+		timeToSend = m.startTime.Add(m.durations[m.index])
+	} else {
+		timeToSend = time.Now()
+	}
+	m.index += 1
+	return timeToSend
+}
+
+// Returns the number of times that the Now function was called
+func (m *mockClock) getIndex() int {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	return m.index
+}
+
+// TestCustomNodeLister can be used to mock nodeLister Get call when testing delayed tainting
+type TestCustomNodeLister struct {
+	nodes                    map[string]*apiv1.Node
+	getCallCount             map[string]int
+	nodeTaintAfterNthGetCall map[string]int
+}
+
+// List returns all nodes in test lister.
+func (l *TestCustomNodeLister) List() ([]*apiv1.Node, error) {
+	var nodes []*apiv1.Node
+	for _, node := range l.nodes {
+		nodes = append(nodes, node)
+	}
+	return nodes, nil
+}
+
+// Get returns node from test lister. Add ToBeDeletedTaint to the node
+// during the N th call specified in the nodeTaintAfterNthGetCall
+func (l *TestCustomNodeLister) Get(name string) (*apiv1.Node, error) {
+	for _, node := range l.nodes {
+		l.getCallCount[node.Name] += 1
```

Comment:
Why are you bumping call count for the node before checking whether it is the one requested?

Reply:
Changed